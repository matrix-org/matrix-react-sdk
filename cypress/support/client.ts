/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import type {
    MatrixClient,
    Room,
    IContent,
    FileType,
    Upload,
    UploadOpts,
    ICreateRoomOpts,
    ISendEventResponse,
} from "matrix-js-sdk/src/matrix";
import Chainable = Cypress.Chainable;

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the MatrixClient from the MatrixClientPeg
             */
            getClient(): Chainable<MatrixClient | undefined>;
            /**
             * Create a room with given options.
             * @param options the options to apply when creating the room
             * @return the ID of the newly created room
             */
            createRoom(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Create a space with given options.
             * @param options the options to apply when creating the space
             * @return the ID of the newly created space (room)
             */
            createSpace(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Invites the given user to the given room.
             * @param roomId the id of the room to invite to
             * @param userId the id of the user to invite
             */
            inviteUser(roomId: string, userId: string): Chainable<{}>;
            /**
             * Sets account data for the user.
             * @param type The type of account data.
             * @param data The data to store.
             */
            setAccountData(type: string, data: object): Chainable<{}>;
            /**
             * @param {string} roomId
             * @param {string} threadId
             * @param {string} eventType
             * @param {Object} content
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            sendEvent(
                roomId: string,
                threadId: string | null,
                eventType: string,
                content: IContent,
            ): Chainable<ISendEventResponse>;
            /**
             * @param {string} name
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setDisplayName(name: string): Chainable<{}>;
            /**
             * @param {string} url
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setAvatarUrl(url: string): Chainable<{}>;
            /**
             * Upload a file to the media repository on the homeserver.
             *
             * @param {object} file The object to upload. On a browser, something that
             *   can be sent to XMLHttpRequest.send (typically a File).  Under node.js,
             *   a a Buffer, String or ReadStream.
             */
            uploadContent(file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>>;
            /**
             * Turn an MXC URL into an HTTP one. <strong>This method is experimental and
             * may change.</strong>
             * @param {string} mxcUrl The MXC URL
             * @param {Number} width The desired width of the thumbnail.
             * @param {Number} height The desired height of the thumbnail.
             * @param {string} resizeMethod The thumbnail resize method to use, either
             * "crop" or "scale".
             * @param {Boolean} allowDirectLinks If true, return any non-mxc URLs
             * directly. Fetching such URLs will leak information about the user to
             * anyone they share a room with. If false, will return null for such URLs.
             * @return {?string} the avatar URL or null.
             */
            mxcUrlToHttp(
                mxcUrl: string,
                width?: number,
                height?: number,
                resizeMethod?: string,
                allowDirectLinks?: boolean,
            ): string | null;
            /**
             * Gets the list of DMs with a given user
             * @param userId The ID of the user
             * @return the list of DMs with that user
             */
            getDmRooms(userId: string): Chainable<string[]>;
            /**
             * Joins the given room by alias or ID
             * @param roomIdOrAlias the id or alias of the room to join
             */
            joinRoom(roomIdOrAlias: string): Chainable<Room>;
        }
    }
}

Cypress.Commands.add("getClient", (): Chainable<MatrixClient | undefined> => {
    return cy.window({ log: false }).then((win) => win.mxMatrixClientPeg.matrixClient);
});

Cypress.Commands.add("getDmRooms", (userId: string): Chainable<string[]> => {
    return cy
        .getClient()
        .then((cli) => cli.getAccountData("m.direct")?.getContent<Record<string, string[]>>())
        .then((dmRoomMap) => dmRoomMap[userId] ?? []);
});

Cypress.Commands.add("createRoom", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.window({ log: false }).then(async (win) => {
        const cli = win.mxMatrixClientPeg.matrixClient;
        const resp = await cli.createRoom(options);
        const roomId = resp.room_id;

        if (!cli.getRoom(roomId)) {
            await new Promise<void>((resolve) => {
                const onRoom = (room: Room) => {
                    if (room.roomId === roomId) {
                        cli.off(win.matrixcs.ClientEvent.Room, onRoom);
                        resolve();
                    }
                };
                cli.on(win.matrixcs.ClientEvent.Room, onRoom);
            });
        }

        return roomId;
    });
});

Cypress.Commands.add("createSpace", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.createRoom({
        ...options,
        creation_content: {
            type: "m.space",
        },
    });
});

Cypress.Commands.add("inviteUser", (roomId: string, userId: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        const res = await cli.invite(roomId, userId);
        Cypress.log({ name: "inviteUser", message: `sent invite in ${roomId} for ${userId}` });
        return res;
    });
});

Cypress.Commands.add("setAccountData", (type: string, data: object): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAccountData(type, data);
    });
});

Cypress.Commands.add(
    "sendEvent",
    (roomId: string, threadId: string | null, eventType: string, content: IContent): Chainable<ISendEventResponse> => {
        return cy.getClient().then(async (cli: MatrixClient) => {
            return cli.sendEvent(roomId, threadId, eventType, content);
        });
    },
);

Cypress.Commands.add("setDisplayName", (name: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setDisplayName(name);
    });
});

Cypress.Commands.add("uploadContent", (file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.uploadContent(file, opts);
    });
});

Cypress.Commands.add("setAvatarUrl", (url: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAvatarUrl(url);
    });
});

Cypress.Commands.add("joinRoom", (roomIdOrAlias: string): Chainable<Room> => {
    return cy.getClient().then((cli) => cli.joinRoom(roomIdOrAlias));
});
