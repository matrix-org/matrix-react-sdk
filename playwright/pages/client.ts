/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { JSHandle, Page } from "@playwright/test";
import { PageFunctionOn } from "playwright-core/types/structs";

import type {
    IContent,
    ICreateRoomOpts,
    ISendEventResponse,
    MatrixClient,
    Room,
    IRoomDirectoryOptions,
} from "matrix-js-sdk/src/matrix";

export class Client {
    protected client: JSHandle<MatrixClient>;

    protected getClientHandle(): Promise<JSHandle<MatrixClient>> {
        return this.page.evaluateHandle(() => window.mxMatrixClientPeg.get());
    }

    public async prepareClient(): Promise<JSHandle<MatrixClient>> {
        if (!this.client) {
            this.client = await this.getClientHandle();
        }
        return this.client;
    }

    public constructor(protected readonly page: Page) {
        page.on("framenavigated", async () => {
            this.client = null;
        });
    }

    public evaluate<R, Arg, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, Arg, R>,
        arg: Arg,
    ): Promise<R>;
    public evaluate<R, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, void, R>,
        arg?: any,
    ): Promise<R>;
    public async evaluate<T>(fn: (client: MatrixClient) => T, arg?: any): Promise<T> {
        await this.prepareClient();
        return this.client.evaluate(fn, arg);
    }

    /**
     * @param roomId ID of the room to send the event into
     * @param threadId ID of the thread to send into or null for main timeline
     * @param eventType type of event to send
     * @param content the event content to send
     */
    public async sendEvent(
        roomId: string,
        threadId: string | null,
        eventType: string,
        content: IContent,
    ): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            async (client, { roomId, threadId, eventType, content }) => {
                return client.sendEvent(roomId, threadId, eventType, content);
            },
            { roomId, threadId, eventType, content },
        );
    }

    /**
     * Send a message as a bot into a room
     * @param roomId ID of the room to send the message into
     * @param content the event content to send
     */
    public async sendMessage(roomId: string, content: IContent): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            (client, { roomId, content }) => {
                return client.sendMessage(roomId, content);
            },
            {
                roomId,
                content,
            },
        );
    }

    /**
     * Create a room with given options.
     * @param options the options to apply when creating the room
     * @return the ID of the newly created room
     */
    public async createRoom(options: ICreateRoomOpts): Promise<string> {
        const client = await this.prepareClient();
        return await client.evaluate(async (cli, options) => {
            const resp = await cli.createRoom(options);
            const roomId = resp.room_id;
            if (!cli.getRoom(roomId)) {
                await new Promise<void>((resolve) => {
                    const onRoom = (room: Room) => {
                        if (room.roomId === roomId) {
                            cli.off(window.matrixcs.ClientEvent.Room, onRoom);
                            resolve();
                        }
                    };
                    cli.on(window.matrixcs.ClientEvent.Room, onRoom);
                });
            }
            return roomId;
        }, options);
    }

    /**
     * Create a space with given options.
     * @param options the options to apply when creating the space
     * @return the ID of the newly created space (room)
     */
    public async createSpace(options: ICreateRoomOpts): Promise<string> {
        return this.createRoom({
            ...options,
            creation_content: {
                ...options.creation_content,
                type: "m.space",
            },
        });
    }

    /**
     * Joins the given room by alias or ID
     * @param roomIdOrAlias the id or alias of the room to join
     */
    public async joinRoom(roomIdOrAlias: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(async (client, roomIdOrAlias) => {
            return await client.joinRoom(roomIdOrAlias);
        }, roomIdOrAlias);
    }

    /**
     * Make this bot join a room by name
     * @param roomName Name of the room to join
     */
    public async joinRoomByName(roomName: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(
            (client, { roomName }) => {
                const room = client.getRooms().find((r) => r.getDefaultRoomName(client.getUserId()) === roomName);
                if (room) {
                    return client.joinRoom(room.roomId);
                }
                throw new Error(`Bot room join failed. Cannot find room '${roomName}'`);
            },
            {
                roomName,
            },
        );
    }

    /**
     * Invites the given user to the given room.
     * @param roomId the id of the room to invite to
     * @param userId the id of the user to invite
     */
    public async inviteUser(roomId: string, userId: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate((client, { roomId, userId }) => client.invite(roomId, userId), {
            roomId,
            userId,
        });
    }

    public async publicRooms(options?: IRoomDirectoryOptions): ReturnType<MatrixClient["publicRooms"]> {
        const client = await this.prepareClient();
        return await client.evaluate((client, options) => {
            return client.publicRooms(options);
        }, options);
    }
}
