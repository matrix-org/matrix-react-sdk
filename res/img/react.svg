<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
