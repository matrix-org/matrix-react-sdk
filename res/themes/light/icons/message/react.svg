<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>