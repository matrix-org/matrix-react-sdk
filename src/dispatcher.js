/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

'use strict';

import flux from "flux";

class MatrixDispatcher extends flux.Dispatcher {
    /**
     * @param {Object|function} payload Required. The payload to dispatch.
     *        If an Object, must contain at least an 'action' key.
     *        If a function, must have the signature (dispatch) => {...}.
     * @param {boolean=} sync Optional. Pass true to dispatch
     *        synchronously. This is useful for anything triggering
     *        an operation that the browser requires user interaction
     *        for.
     */
    dispatch(payload, sync) {
        // Allow for asynchronous dispatching by accepting payloads that have the
        // type `function (dispatch) {...}`
        if (typeof payload === 'function') {
            payload((action) => {
                this.dispatch(action, sync);
            });
            return;
        }

        if (sync) {
            super.dispatch(payload);
        } else {
            // Unless the caller explicitly asked for us to dispatch synchronously,
            // we always set a timeout to do this: The flux dispatcher complains
            // if you dispatch from within a dispatch, so rather than action
            // handlers having to worry about not calling anything that might
            // then dispatch, we just do dispatches asynchronously.
            setTimeout(super.dispatch.bind(this, payload), 0);
        }
    }
}

if (global.mxDispatcher === undefined) {
    global.mxDispatcher = new MatrixDispatcher();
}
export default global.mxDispatcher;
