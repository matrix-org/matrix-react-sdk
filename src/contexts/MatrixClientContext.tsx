/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, {
    ComponentClass,
    createContext,
    forwardRef,
    PropsWithoutRef,
    ForwardRefExoticComponent,
    useContext,
    RefAttributes,
} from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";

// This context is available to components under LoggedInView,
// the context must not be used by components outside a MatrixClientContext tree.
// This assertion allows us to make the type not nullable.
const MatrixClientContext = createContext<MatrixClient>(null as any);
MatrixClientContext.displayName = "MatrixClientContext";
export default MatrixClientContext;

export interface MatrixClientProps {
    mxClient: MatrixClient;
}

export function useMatrixClientContext(): MatrixClient {
    return useContext(MatrixClientContext);
}

const matrixHOC = <ComposedComponentProps extends {}>(
    ComposedComponent: ComponentClass<ComposedComponentProps>,
): ForwardRefExoticComponent<
    PropsWithoutRef<Omit<ComposedComponentProps, "mxClient">> &
        RefAttributes<InstanceType<ComponentClass<ComposedComponentProps>>>
> => {
    type ComposedComponentInstance = InstanceType<typeof ComposedComponent>;

    // eslint-disable-next-line react-hooks/rules-of-hooks

    const TypedComponent = ComposedComponent;

    return forwardRef<ComposedComponentInstance, Omit<ComposedComponentProps, "mxClient">>((props, ref) => {
        const client = useContext(MatrixClientContext);

        // @ts-ignore
        return <TypedComponent ref={ref} {...props} mxClient={client} />;
    });
};
export const withMatrixClientHOC = matrixHOC;
