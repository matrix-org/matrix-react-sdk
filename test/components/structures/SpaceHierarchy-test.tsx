/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render } from "@testing-library/react";
import { MatrixClient } from "matrix-js-sdk/src/client";
import { Room } from "matrix-js-sdk/src/models/room";
import { RoomHierarchy } from "matrix-js-sdk/src/room-hierarchy";
import { IHierarchyRoom } from "matrix-js-sdk/src/@types/spaces";

import { MatrixClientPeg } from "../../../src/MatrixClientPeg";
import { mkStubRoom, stubClient } from "../../test-utils";
import dispatcher from "../../../src/dispatcher/dispatcher";
import { HierarchyLevel, showRoom, toLocalRoom } from "../../../src/components/structures/SpaceHierarchy";
import { Action } from "../../../src/dispatcher/actions";
import MatrixClientContext from "../../../src/contexts/MatrixClientContext";
import DMRoomMap from "../../../src/utils/DMRoomMap";

// Fake random strings to give a predictable snapshot for checkbox IDs
jest.mock("matrix-js-sdk/src/randomstring", () => {
    return {
        randomString: () => "abdefghi",
    };
});

describe("SpaceHierarchy", () => {
    describe("showRoom", () => {
        let client: MatrixClient;
        let hierarchy: RoomHierarchy;
        let room: Room;
        beforeEach(() => {
            stubClient();
            client = MatrixClientPeg.get();
            room = new Room("room-id", client, "@alice:example.com");
            hierarchy = new RoomHierarchy(room);

            jest.spyOn(client, "isGuest").mockReturnValue(false);

            jest.spyOn(hierarchy.roomMap, "get").mockReturnValue({
                children_state: [],
                room_id: "room-id2",
                canonical_alias: "canonical-alias",
                aliases: ["uncanonical-alias", "canonical-alias"],
                world_readable: true,
                guest_can_join: false,
                num_joined_members: 35,
            });

            jest.spyOn(dispatcher, "dispatch");
        });

        it("shows room", () => {
            showRoom(client, hierarchy, "room-id2");
            expect(dispatcher.dispatch).toHaveBeenCalledWith({
                action: Action.ViewRoom,
                should_peek: true,
                room_alias: "canonical-alias",
                room_id: "room-id2",
                via_servers: [],
                oob_data: {
                    avatarUrl: undefined,
                    name: "canonical-alias",
                },
                roomType: undefined,
                metricsTrigger: "RoomDirectory",
            });
        });
    });

    describe("toLocalRoom", () => {
        stubClient();
        const client = MatrixClientPeg.get();
        const roomV1 = mkStubRoom("room-id-1", "Room V1", client);
        const roomV2 = mkStubRoom("room-id-2", "Room V2", client);
        const roomV3 = mkStubRoom("room-id-3", "Room V3", client);
        jest.spyOn(client, "getRoomUpgradeHistory").mockReturnValue([roomV1, roomV2, roomV3]);

        it("grabs last room that is in hierarchy when latest version is in hierarchy", () => {
            const hierarchy = {
                roomMap: new Map([
                    [roomV1.roomId, { room_id: roomV1.roomId } as IHierarchyRoom],
                    [roomV2.roomId, { room_id: roomV2.roomId } as IHierarchyRoom],
                    [roomV3.roomId, { room_id: roomV3.roomId } as IHierarchyRoom],
                ]),
            } as RoomHierarchy;
            const localRoomV1 = toLocalRoom(client, { room_id: roomV1.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV1.room_id).toEqual(roomV3.roomId);
            const localRoomV2 = toLocalRoom(client, { room_id: roomV2.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV2.room_id).toEqual(roomV3.roomId);
            const localRoomV3 = toLocalRoom(client, { room_id: roomV3.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV3.room_id).toEqual(roomV3.roomId);
        });

        it("grabs last room that is in hierarchy when latest version is *not* in hierarchy", () => {
            const hierarchy = {
                roomMap: new Map([
                    [roomV1.roomId, { room_id: roomV1.roomId } as IHierarchyRoom],
                    [roomV2.roomId, { room_id: roomV2.roomId } as IHierarchyRoom],
                ]),
            } as RoomHierarchy;
            const localRoomV1 = toLocalRoom(client, { room_id: roomV1.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV1.room_id).toEqual(roomV2.roomId);
            const localRoomV2 = toLocalRoom(client, { room_id: roomV2.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV2.room_id).toEqual(roomV2.roomId);
            const localRoomV3 = toLocalRoom(client, { room_id: roomV3.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV3.room_id).toEqual(roomV2.roomId);
        });

        it("returns specified room when none of the versions is in hierarchy", () => {
            const hierarchy = { roomMap: new Map([]) } as RoomHierarchy;
            const localRoomV1 = toLocalRoom(client, { room_id: roomV1.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV1.room_id).toEqual(roomV1.roomId);
            const localRoomV2 = toLocalRoom(client, { room_id: roomV2.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV2.room_id).toEqual(roomV2.roomId);
            const localRoomV3 = toLocalRoom(client, { room_id: roomV3.roomId } as IHierarchyRoom, hierarchy);
            expect(localRoomV3.room_id).toEqual(roomV3.roomId);
        });
    });

    describe("<HierarchyLevel />", () => {
        stubClient();
        const client = MatrixClientPeg.get();

        const dmRoomMap = {
            getUserIdForRoomId: jest.fn(),
        } as unknown as DMRoomMap;
        jest.spyOn(DMRoomMap, "shared").mockReturnValue(dmRoomMap);

        const root = mkStubRoom("room-id-1", "Room 1", client);
        const room1 = mkStubRoom("room-id-2", "Room 2", client);
        const room2 = mkStubRoom("room-id-3", "Room 3", client);

        const hierarchyRoot = {
            room_id: root.roomId,
            num_joined_members: 1,
            children_state: [
                {
                    state_key: room1.roomId,
                    content: { order: "1" },
                },
                {
                    state_key: room2.roomId,
                    content: { order: "2" },
                },
            ],
        } as IHierarchyRoom;
        const hierarchyRoom1 = { room_id: room1.roomId, num_joined_members: 2 } as IHierarchyRoom;
        const hierarchyRoom2 = { room_id: root.roomId, num_joined_members: 3 } as IHierarchyRoom;

        const roomHierarchy = {
            roomMap: new Map([
                [root.roomId, hierarchyRoot],
                [room1.roomId, hierarchyRoom1],
                [room2.roomId, hierarchyRoom2],
            ]),
            isSuggested: jest.fn(),
        } as unknown as RoomHierarchy;

        it("renders", () => {
            const defaultProps = {
                root: hierarchyRoot,
                roomSet: new Set([hierarchyRoom1, hierarchyRoom2]),
                hierarchy: roomHierarchy,
                parents: new Set<string>(),
                selectedMap: new Map<string, Set<string>>(),
                onViewRoomClick: jest.fn(),
                onJoinRoomClick: jest.fn(),
                onToggleClick: jest.fn(),
            };
            const getComponent = (props = {}): React.ReactElement => (
                <MatrixClientContext.Provider value={client}>
                    <HierarchyLevel {...defaultProps} {...props} />;
                </MatrixClientContext.Provider>
            );

            const { container } = render(getComponent());
            expect(container).toMatchSnapshot();
        });
    });
});
