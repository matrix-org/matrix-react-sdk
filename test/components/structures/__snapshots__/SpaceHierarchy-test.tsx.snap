// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceHierarchy <HierarchyLevel /> renders 1`] = `
<div>
  <li
    class="mx_SpaceHierarchy_roomTileWrapper"
    role="treeitem"
  >
    <div
      class="mx_AccessibleButton mx_SpaceHierarchy_roomTile"
      role="button"
      tabindex="-1"
    >
      <div
        class="mx_SpaceHierarchy_roomTile_item"
      >
        <div
          class="mx_SpaceHierarchy_roomTile_avatar"
        >
          <img
            alt=""
            class="mx_BaseAvatar mx_BaseAvatar_image"
            data-testid="avatar-img"
            src="http://this.is.a.url/avatar.url/room.png"
            style="width: 20px; height: 20px;"
          />
        </div>
        <div
          class="mx_SpaceHierarchy_roomTile_name"
        >
          Unnamed Room
          <div
            class="mx_SpaceHierarchy_roomTile_joined"
          >
            Joined
          </div>
        </div>
        <div
          class="mx_SpaceHierarchy_roomTile_info"
        >
          2 members
           · 
          <span
            dir="auto"
          />
        </div>
      </div>
      <div
        class="mx_SpaceHierarchy_actions"
      >
        <div
          class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
          role="button"
          tabindex="-1"
        >
          View
        </div>
        <span
          class="mx_Checkbox mx_Checkbox_hasKind mx_Checkbox_kind_solid"
        >
          <input
            id="checkbox_abdefghi"
            tabindex="-1"
            type="checkbox"
          />
          <label
            for="checkbox_abdefghi"
          >
            <div
              class="mx_Checkbox_background"
            >
              <div
                class="mx_Checkbox_checkmark"
              />
            </div>
          </label>
        </span>
      </div>
    </div>
  </li>
  <li
    class="mx_SpaceHierarchy_roomTileWrapper"
    role="treeitem"
  >
    <div
      class="mx_AccessibleButton mx_SpaceHierarchy_roomTile"
      role="button"
      tabindex="-1"
    >
      <div
        class="mx_SpaceHierarchy_roomTile_item"
      >
        <div
          class="mx_SpaceHierarchy_roomTile_avatar"
        >
          <img
            alt=""
            class="mx_BaseAvatar mx_BaseAvatar_image"
            data-testid="avatar-img"
            src="http://this.is.a.url/avatar.url/room.png"
            style="width: 20px; height: 20px;"
          />
        </div>
        <div
          class="mx_SpaceHierarchy_roomTile_name"
        >
          Unnamed Room
          <div
            class="mx_SpaceHierarchy_roomTile_joined"
          >
            Joined
          </div>
        </div>
        <div
          class="mx_SpaceHierarchy_roomTile_info"
        >
          3 members
           · 
          <span
            dir="auto"
          />
        </div>
      </div>
      <div
        class="mx_SpaceHierarchy_actions"
      >
        <div
          class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
          role="button"
          tabindex="-1"
        >
          View
        </div>
        <span
          class="mx_Checkbox mx_Checkbox_hasKind mx_Checkbox_kind_solid"
        >
          <input
            id="checkbox_abdefghi"
            tabindex="-1"
            type="checkbox"
          />
          <label
            for="checkbox_abdefghi"
          >
            <div
              class="mx_Checkbox_background"
            >
              <div
                class="mx_Checkbox_checkmark"
              />
            </div>
          </label>
        </span>
      </div>
    </div>
  </li>
  ;
</div>
`;
